src/store.rs

```diff
@@ -8,130 +9,159 @@ use zkevm_opcode_defs::{
-#[derive(Debug, Clone, Default)]
-pub struct InMemory {
-    contract_storage: HashMap<U256, Vec<U256>>,
-    pub state_storage: HashMap<StorageKey, U256>,
+impl InitialStorage for InitialStorageMemory {
+    fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
+        Ok(self.initial_storage.get(&key).copied())
+    }
 }
 
-#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
-pub struct StorageKey {
-    pub address: H160,
-    pub key: U256,
+pub trait ContractStorage: Debug {
+    fn decommit(&self, hash: U256) -> Result<Option<Vec<U256>>, StorageError>;
+}
+#[derive(Debug)]
+pub struct ContractStorageMemory {
+    pub contract_storage: HashMap<U256, Vec<U256>>,
 }
 
-impl StorageKey {
-    pub fn new(address: H160, key: U256) -> Self {
-        Self { address, key }
+impl ContractStorage for ContractStorageMemory {
+    fn decommit(&self, hash: U256) -> Result<Option<Vec<U256>>, StorageError> {
+        Ok(self.contract_storage.get(&hash).cloned())
     }
 }
 
-impl InMemory {
-    pub fn new_empty() -> Self {
-        let state_storage = HashMap::new();
-        let contract_storage = HashMap::new();
-        InMemory {
-            state_storage,
-            contract_storage,
-        }
-    }
+#[derive(Debug)]
+pub struct StateStorage {
+    pub storage_changes: HashMap<StorageKey, U256>,
+    pub initial_storage: Rc<RefCell<dyn InitialStorage>>,
+    l1_to_l2_logs: Vec<L2ToL1Log>,
+}
 
-    pub fn new(
-        contract_storage: HashMap<U256, Vec<U256>>,
-        state_storage: HashMap<StorageKey, U256>,
-    ) -> Self {
+impl Default for StateStorage {
+    fn default() -> Self {
         Self {
-            contract_storage,
-            state_storage,
+            storage_changes: HashMap::new(),
+            initial_storage: Rc::new(RefCell::new(InitialStorageMemory {
+                initial_storage: HashMap::new(),
+            })),
+            l1_to_l2_logs: Vec::new(),
         }
     }
 }
 
-impl Storage for InMemory {
-    fn decommit(&self, hash: U256) -> Result<Option<Vec<U256>>, StorageError> {
-        Ok(self.contract_storage.get(&hash).cloned())
-    }
-
-    fn add_contract(&mut self, hash: U256, code: Vec<U256>) -> Result<(), StorageError> {
-        self.contract_storage.insert(hash, code);
-        Ok(())
+impl StateStorage {
+    pub fn new(initial_storage: Rc<RefCell<dyn InitialStorage>>) -> Self {
+        Self {
+            storage_changes: HashMap::new(),
+            initial_storage,
+            l1_to_l2_logs: Vec::new(),
+        }
     }
 
-    fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
-        Ok(self.state_storage.get(&key).copied())
+    pub fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
+        let mut value = self.storage_changes.get(&key).copied();
+        if value.is_none() {
+            value = self.initial_storage.borrow().storage_read(key)?;
+        }
+        Ok(value)
     }
```

Comment:
Nit-pick: an optional style suggestion, up to you if you want to apply this change 😄 .
```suggestion
        pub fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
        match self.storage_changes.get(&key) {
            None => self.initial_storage.borrow().storage_read(key),
            value => Ok(value.copied()),
        }
    }
```